WalletWasabi/Crypto/GroupElement.cs

```diff
@@ -59,5 +60,24 @@ public GroupElement(GEJ groupElement)
 		}
 
 		public static bool operator !=(GroupElement a, GroupElement b) => !(a == b);
+
+		public Scalar Sha256(GroupElement groupElement)
```

Comment:
wait, i think I misread the code, I thought it was a static method that constructs scalars from group elements on first read, but this appears to be an instance method of the encapsulated group element that also takes *another* group element as an argument... why is that?

Reply:
Yes, @nothingmuch is right here but I would go further and say that a group element has not "conceptually" a sha256.

Reply:
```cs
	public static class ZkChallenge
	{
		public static Scalar Build(GE publicPoint, GE randomPoint)
		{
			Guard.True($"{nameof(publicPoint)}.{nameof(publicPoint.IsValidVariable)}", publicPoint.IsValidVariable);
			Guard.True($"{nameof(randomPoint)}.{nameof(randomPoint.IsValidVariable)}", randomPoint.IsValidVariable);
			if (Secp256k1Helpers.Equals(publicPoint, randomPoint))
			{
				throw new InvalidOperationException($"{nameof(publicPoint)} and {nameof(randomPoint)} should not be equal.");
			}

			var concatenation = ByteHelpers.Combine(
				publicPoint.x.ToBytes(),
				publicPoint.y.ToBytes(),
				randomPoint.x.ToBytes(),
				randomPoint.y.ToBytes());
			using var sha256 = System.Security.Cryptography.SHA256.Create();
			var hash = sha256.ComputeHash(concatenation);
			var challenge = new Scalar(hash);
			return challenge;
		}
	}
```

https://github.com/zkSNACKs/WalletWasabi/pull/3915/files#diff-94391fcf9b0bdadb1f101ea58c077b6eR7-R30

Reply:
I think that's cleaner but still only works for proofs which require a single public input and a single challenge, e.g. Schnorr identification (knowledge of exponent), which is not general enough for the other statements

Reply:
look how genzkp is used here: https://github.com/gdanezis/petlib/blob/master/examples/amacscreds.py

or Merlin transcripts: https://merlin.cool/

Reply:
This is how I did it:

https://github.com/lontivero/Wabisabi/blob/9cd24f8c175acf97c3f9c49b077a8714af178745/Wabisabi/Crypto.cs#L226-L240

This is how the petlib project does it:

https://github.com/gdanezis/petlib/blob/d69649d9079f64bfc65134cd3594c248de6ce67e/examples/zkp.py#L10-L18

Both are almost identical except that the petlib converts the elements to strings and concatenates the length as prefix, what makes not a lot of sense to me.